3.8 Docs (#11102)
apollographql/apollo-client: docs/source/api/link/apollo-link-remove-typename.mdx

---
title: Remove Typename Link
description: Automatically remove __typename fields from variables.
minVersion: 3.8.0
---

## Overview

The `removeTypenameFromVariables` link automatically removes `__typename` fields from variables in operations. This is useful when reusing data from a query as an argument to another GraphQL operation.

```ts
import { removeTypenameFromVariables } from '@apollo/client/link/remove-typename';

const removeTypenameLink = removeTypenameFromVariables();
```

> 

Reviewer: There's a header for keeping typename, but none for the example of removing it.

```suggestion

## Remove `__typename` from all variables 
```

Author: Resolved in https://github.com/apollographql/apollo-client/pull/11136/commits/4d7f7f834b3fa54e3a1a60e54ffc1bd434bb2bca

> As an example, let's take the following query. Apollo Client will automatically add `__typename` fields to the query for each field selection set.

Author: Resolved in https://github.com/apollographql/apollo-client/pull/11136/commits/44d6d02a2f2ac98ffd8738c4e5f24f1e5f691d6d.

```ts
const query = gql`
  query DashboardQuery($id: ID!) {
    dashboard(id: $id) {
      id
      name
    }
  }
`;

const { data } = await client.query({ query, variables: { id: 1 }});
// {
//   "dashboard": {
//     "__typename": "Dashboard",
//     "id": 1,
//     "name": "My Dashboard"
//   }
// }
```

Now let's update this dashboard by sending a mutation to our server. We'll use the `dashboard` as input to our mutation.

```ts
const mutation = gql`
  mutation UpdateDashboardMutation($dashboard: DashboardInput!) {
    updateDashboard(dashboard: $dashboard) {
      id
      name
    }
  }
`;

await client.mutate({ 
  mutation,
  variables: { 
    dashboard: { ...data.dashboard, name: 'My Updated Dashboard' } 
  }
});
```

Without the use of the `removeTypenameFromVariables` link, the server will return an error because `data.dashboard` still contains the `__typename` field.

## Keep `__typename` in JSON scalars

> While the [GraphQL spec](https://spec.graphql.org/October2021/#sec-Input-Objects.Type-Validation) disallows input fields that begin with the characters "\_\_" (two underscores), this restriction is not applied when the input field is a JSON scalar (i.e. a scalar type that accepts raw JSON as input). When your server relies on `__typename` within certain `JSON` scalars, you can configure the `removeTypenameFromVariables` link to retain `__typename`.

Author: Resolved in https://github.com/apollographql/apollo-client/pull/11136/commits/129e377fc324d65d3622502f64925b1481feec88

> > Note: the JSON scalar type does not need to be literally named `JSON` to be considered a JSON scalar.

Author: Resolved in https://github.com/apollographql/apollo-client/pull/11136/commits/5dd50ed54abfef430e357df0516e5eb85d06c69e

> Provide an `except` option to map all variable types that should maintain `__typename`. This is done using the `KEEP` sentinel. Each key in the `except` option should correspond to an input type in your GraphQL schema.

Reviewer: ```suggestion
To do so, provide an `except` option when instantiating `removeTypenameFromVariables` and use the `KEEP` sentinel to denote which variables types should keep `__typename`. Each key in the `except` option should correspond to an input type in your GraphQL schema.
```

Author: Resolved in https://github.com/apollographql/apollo-client/pull/11136/commits/23826ef25f13b505b60141822a03a6618e1009f9

> Here we tell the `removeTypenameFromVariables` link to keep all `__typename` fields for any variable that is declared as a `JSON` type. Variable types are inferred from the GraphQL query. Let's take the following GraphQL mutation:

Reviewer: It would be helpful to start with the context of this schema:

```suggestion
For example, suppose your schema includes a `ConfigureDashboardMutation` mutation that takes a `JSON` type variable named `$dashboardConfig`:
```

Author: Resolved in https://github.com/apollographql/apollo-client/pull/11136/commits/8616fd721040de3636a1e896c425850ae5a020bd

```graphql
mutation ConfigureDashboardMutation($dashboardConfig: JSON) {
  configureDashboard(config: $dashboardConfig) {
    id
  }
}
```

> 

Reviewer: ```suggestion

You can tell the `removeTypenameFromVariables` link to keep all `__typename` fields for any variable declared as a `JSON` type. (Variable types are inferred from the GraphQL query.) 
```

Author: Resolved in https://github.com/apollographql/apollo-client/pull/11136/commits/31bbe61c7e55cd670b8b1919e5a7332d7fe827c3

```ts
import { removeTypenameFromVariables, KEEP } from '@apollo/client/link/remove-typename';

const removeTypenameLink = removeTypenameFromVariables({
  except: {
    JSON: KEEP
  }
});
```

> Here we've declared a `dashboardConfig` variable as type `JSON`. When the query moves through the `removeTypenameFromVariables` link, the `dashboardConfig` variable will be detected as a `JSON` scalar type and all `__typename` fields are kept intact.

Reviewer: ```suggestion
When the query moves through the `removeTypenameFromVariables` link, the `dashboardConfig` variable will be detected as a `JSON` scalar type and all `__typename` fields are kept intact.
```

Reviewer: Might it be helpful to include a sample query and response?

Author: Resolved the first comment in https://github.com/apollographql/apollo-client/pull/11136/commits/6387bf99b31589434374a83771a4014fd3ad6048

Will leave this thread unresolved re: question

### Nested JSON scalar fields in input variables

> Not all top-level variables may map to a JSON scalar type. For more complex input object types, the JSON scalar may be found on a nested field. The `except` option allows you to configured nested fields within these types to keep `__typename` intact for these fields.

Reviewer: ```suggestion
Not all top-level variables may map to a JSON scalar type. For more complex input object types, the JSON scalar may be found on a nested field. The `except` option lets you to configure nested fields within these types to keep `__typename` intact.
```

Author: Resolved in https://github.com/apollographql/apollo-client/pull/11136/commits/131af796bbcc4ca77b65a0e8892bf7b489d54c0d
